File: composer/modules/web/src/core/editor/model/Editor.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

import EventChannel from 'event_channel';
import { EVENTS } from './../constants';
import UndoManager from './../undo-manager/undo-manager';

/**
 * Class to represent an editor tab
 */
class Editor extends EventChannel {

    /**
     * Creates an editor tab
     * @param {File} file Target file
     * @param {Object} definition Editor Definition
     */
    constructor(file, definition) {
        super();
        this._file = file;
        this._definition = definition;
        this._customTitleClass = definition.tabTitleClass || '';
        this._undoManager = new UndoManager();
        this._props = {};
    }

    /**
     * Gets current undo manager
     *
     * @returns {UndoManager} current undo manager
     */
    get undoManager() {
        return this._undoManager;
    }

    /**
     * Sets undo manager
     *
     * @param {UndoManager} undoManager
     */
    set undoManager(undoManager) {
        this._undoManager = undoManager;
    }


    /**
     * Returns id
     */
    get id() {
        return this._file.fullPath;
    }

    /**
     * Returns file
     */
    get file() {
        return this._file;
    }

    /**
     * Sets file
     */
    set file(file) {
        this._file = file;
    }

    /**
     * Returns isDirty
     */
    get isDirty() {
        return this._file.isDirty;
    }

    /**
     * Returns editor definition
     */
    get definition() {
        return this._definition;
    }

    /**
     * Sets editor definition
     */
    set definition(definition) {
        this._definition = definition;
    }

    /**
     * Returns custom class for title
     */
    get customTitleClass() {
        return this._customTitleClass;
    }

    /**
     * Sets custom class for title
     */
    set customTitleClass(customTitleClass) {
        this._customTitleClass = customTitleClass;
        this.trigger(EVENTS.UPDATE_TAB_TITLE, this);
    }

    /**
     * Get the value of given property.
     *
     * @param {string} propertyName
     */
    getProperty(propertyName) {
        return this._props[propertyName];
    }

    /**
     * Set the value of given property.
     *
     * @param {string} propertyName
     * @param {any} propertyValue
     */
    setProperty(propertyName, propertyValue) {
        this._props[propertyName] = propertyValue;
    }

}

export default Editor;

File: tests/language-server-simulator/src/main/java/org/ballerinalang/langserver/simulator/Editor.java
/*
 * Copyright (c) 2021, WSO2 Inc. (http://wso2.com) All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.ballerinalang.langserver.simulator;

import org.ballerinalang.langserver.BallerinaLanguageServer;
import org.ballerinalang.langserver.util.TestUtil;
import org.eclipse.lsp4j.jsonrpc.Endpoint;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Represents the editor used by the end user. An  editor consists of a set of open tabs.
 *
 * @since 2.0.0
 */
public class Editor {

    private static final Logger logger = LoggerFactory.getLogger(Editor.class);

    private BallerinaLanguageServer languageServer;
    private Endpoint endpoint;

    private final List<EditorTab> tabs = new ArrayList<>();
    private EditorTab activeTab;

    private Editor(BallerinaLanguageServer languageServer, Endpoint endpoint) {
        this.languageServer = languageServer;
        this.endpoint = endpoint;
    }

    /**
     * Simulates opening the editor. Here we initialize the language server.
     *
     * @return Editor instance
     */
    public static Editor open() {
        BallerinaLanguageServer languageServer = new BallerinaLanguageServer();

        EditorOutputStream outputStream = new EditorOutputStream();
        Endpoint endpoint = TestUtil.initializeLanguageSever(languageServer, outputStream);

        Editor editor = new Editor(languageServer, endpoint);
        outputStream.setEditor(editor);
        return editor;
    }

    public EditorTab openFile(Path filePath) {
        EditorTab editorTab = tabs.stream()
                .filter(tab -> tab.filePath().equals(filePath))
                .findFirst()
                .orElseGet(() -> {
                    EditorTab tab = new EditorTab(filePath, endpoint, languageServer);
                    tabs.add(tab);
                    return tab;
                });
        this.activeTab = editorTab;
        return editorTab;
    }

    public void closeFile(Path filePath) {
        tabs.removeIf(tab -> {
            if (filePath.equals(tab.filePath())) {
                tab.close();
                return true;
            }
            return false;
        });
    }

    public void closeTab(EditorTab tab) {
        tabs.remove(tab);
    }

    public void close() {
        this.languageServer.shutdown();
        tabs.forEach(tab -> tab.close());
    }

    public EditorTab activeTab() {
        return activeTab;
    }
}
